### PR TITLE
CIFAR 10 Classifier using Conv Net

### PR DESCRIPTION
**Pull request by Davidportlouis**
- [x] Added Simple CNN architecture for classifying CIFAR 10 (WIP)
- [x] Pretrained Model
- [x] Notebook for visualization & predicting
- [x] Dataset Source

Solved: ~~Encountering this (PFA Screenshot below) issue at the end of epoch 1, any help or suggestions would be helpful~~

Dataset source: https://drive.google.com/file/d/1zTBYb1oPCJEd0wUbkCuXS6xSA1N1QXLM/view?usp=sharing

### CODE REVIEW

File: cifar10_cnn/cifar_train.cpp [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    model.Add<LeakyReLU<>>();
+    model.Add<BatchNorm<>>(128);
+    model.Add<Convolution<>>(128, 128, 3, 3, 1, 1, 1, 1, 8, 8);
+    model.Add<LeakyReLU<>>();
+    model.Add<BatchNorm<>>(128);
+    model.Add<MaxPooling<>>(2, 2, 2, 2, true);
+    model.Add<Dropout<>>(0.4);
+
+
+    model.Add<Linear<>>(4*4*128, 10);
```
Comment (zoq) [inline cifar10_cnn/cifar_train.cpp:73]: Looks like it's a label issue, can you increase this to e.g. 20 and see if that works?

Comment (Davidportlouis) [reply cifar10_cnn/cifar_train.cpp:73]: It was due to header col, begin recognized as extra class

**Comment by zoq** [comment]
Dataset uploaded - http://datasets.mlpack.org/cifar-10.tar.xz

File: cifar10_cnn/cifar_train.cpp [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+using namespace mlpack;
+using namespace mlpack::ann;
+using namespace arma;
+using namespace std;
+using namespace ens;
+
+// Utility function to generate class labels
+// from probabilities.
+arma::Row<size_t> getLabels(arma::mat yPreds)
```
Comment (zoq) [inline cifar10_cnn/cifar_train.cpp:23]:
```suggestion
arma::Row<size_t> getLabels(const arma::ma&t yPreds) 
```

To avoid an extra copy.

File: cifar10_cnn/periodic_save.hpp [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  /**
+   * Constructor for PeriodicSave class.
+   *
+   * @param network Network type which will be saved periodically.
+   * @param filePath Base path / folder where weights will be saved.
+   * @param modelPrefix Weights will be stored as
+   *    modelPrefix_epoch_loss.bin.
+   * @param period Period after which the model will be saved.
+   * @param silent Boolean to determine whether or not to print saving
+   *    of model.
```
Comment (zoq) [inline cifar10_cnn/periodic_save.hpp:39]:
```suggestion
   *      modelPrefix_epoch_loss.bin.
   * @param period Period after which the model will be saved.
   * @param silent Boolean to determine whether or not to print saving
   *      of model.
```

Should be tabbed twice (4 spaces).

File: cifar10_cnn/cifar10_eval.ipynb [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "id": "b37b63e4-561a-4528-8db9-786336ff91cb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "mat testData, testY;\n",
+    "mlpack::data::Load(\"../data/cifar-10_test.csv\", testData, true);\n",
```
Comment (zoq) [inline cifar10_cnn/cifar10_eval.ipynb:133]: With that we have to download the dataset in the notebook as well.

File: cifar10_cnn/cifar10_eval.ipynb [hunk 5]
```diff
@@ -0,0 +1,193 @@
+{
+ "cells": [
+  {
+   "cell_type": "code",
```
Comment (zoq) [inline cifar10_cnn/cifar10_eval.ipynb:4]: Would be a good idea to add some comments to the top to explain what this notebook is about.

File: cifar10_cnn/cifar10_eval.ipynb (hunk 4 above)
Comment (Davidportlouis) [reply cifar10_cnn/cifar10_eval.ipynb:133]: The dateset in the notebook as well as for the training should be downloaded using the script in tools/download_data_set.py using args cifar10, may be it's good idea to call it from jupyter notbook for evaluation purpose.

File: cifar10_cnn/cifar10_eval.ipynb (hunk 5 above)
Comment (Davidportlouis) [reply cifar10_cnn/cifar10_eval.ipynb:4]: I have add some explanations and markdowns in the notebook

File: cifar10_cnn/cifar_train.cpp [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+{
+    arma::Row<size_t> yLabels(yPreds.n_cols);
+    for (arma::uword i = 0; i < yPreds.n_cols; ++i)
+    {
+        yLabels(i) = yPreds.col(i).index_max();
+    }
+    return yLabels;
+}
+
+int main() {
```
Comment (zoq) [inline cifar10_cnn/cifar_train.cpp:33]:
```suggestion
int main()
{
```

File: cifar10_cnn/cifar_train.cpp [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+    // Train the CNN model using Adam Optimizer and above hyperparameters.
+    ens::Adam optimizer(STEP_SIZE, BATCH_SIZE, 0.9, 0.999, 1e-8, numIterations, 1e-8, true);
+    model.Train(trainX,
+            trainY,
+            optimizer,
+            ens::PrintLoss(),
+            ens::ProgressBar(),
+            ens::EarlyStopAtMinLoss(),
+            ens::PeriodicSave<FFN<NegativeLogLikelihood<>, RandomInitialization>>(model, "./new_models/"));
```
Comment (zoq) [inline cifar10_cnn/cifar_train.cpp:111]: I guess you could also use `decltype` here, instead of manually setting the template parameter, might be cleaner.

File: cifar10_cnn/cifar10_eval.ipynb [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+   "cell_type": "markdown",
+   "id": "47129c1e-26e6-4916-b704-0ba00bf2f503",
+   "metadata": {},
+   "source": [
+    "### CIFAR10 Classifer Evaluation\n",
+    "In this notebook, we evaluate pre-trained **CNN** to classify images from the CIFAR-10 database.\n",
+    "\n",
+    "The images in this database are small color images that fall into one of ten classes; some example images are pictured below.\n",
+    "\n",
+    "<img src='notebook_ims/cifar_data.png' width=50% height=50% />"
```
Comment (zoq) [inline cifar10_cnn/cifar10_eval.ipynb:13]: Looks like `cifar_data.png` is not yet part of the PR.

File: cifar10_cnn/cifar_train.cpp (hunk 7 above)
Comment (Davidportlouis) [reply cifar10_cnn/cifar_train.cpp:111]: I have not used decltype, can you point me to some examples.

**Review by zoq — COMMENTED** [review]
I'll get the conda package ready, and then I think we can merge this in.

**Comment by Davidportlouis** [comment]
> I'll get the conda package ready, and then I think we can merge this in.

Sounds good

**Comment by zoq** [comment]
Almost ready, have to find a way to automatically push it to conda.

**Comment by Davidportlouis** [comment]
> Almost ready, have to find a way to automatically push it to conda.

Maybe this will help https://stackoverflow.com/a/66539742

**Comment by zoq** [comment]
Let's see if http://ci.mlpack.org/job/mlpack-conda-nightly/3/ works.

**Comment by zoq** [comment]
Looks like all the packages are uploaded - https://anaconda.org/zoq/mlpack/files let's see if that works.

**Comment by Davidportlouis** [comment]
> Looks like all the packages are uploaded - https://anaconda.org/zoq/mlpack/files let's see if that works.

Sure, I'll check it out

**Comment by Davidportlouis** [comment]
> > Looks like all the packages are uploaded - https://anaconda.org/zoq/mlpack/files let's see if that works.
> 
> Sure, I'll check it out

It's not working as expected, notebook dies or runs predict method infinitely

**Comment by zoq** [comment]
Okay, thanks for the update, I think in this case we have to wait until the ann code restructure code is finished. In any case let's update the Jupyter local setup script, and merge this afterwards, since the code works in a local C++ env it should work in a notebook as well.

**Comment by zoq** [comment]
> Pre-trained weights for CIFAR10 classifier: https://drive.google.com/file/d/1uRE-2NeKeij44J368lSoSJGqPfPV_y8A/view?usp=sharing (test accuracy 46.38%)
> 
> I have also updated comments in notebook regarding issues in model inference & added some standalone C++ file for evaluating the model on test data.

Uploaded http://datasets.mlpack.org/cifarNet.xml.

File: cifar10_cnn/cifar10_eval.ipynb [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ca47e69f-68ef-4d31-b6c2-44a03b8efbbd",
+   "metadata": {
+    "tags": []
+   },
+   "source": [
+    "<h4><font color='red'>Note: Model inference does not work properly in the notebook, the below cell executes forever.</font></h4>\n",
```
Comment (zoq) [inline cifar10_cnn/cifar10_eval.ipynb:209]: Thanks for adding the comment.

Comment (Davidportlouis) [reply cifar10_cnn/cifar10_eval.ipynb:209]: Any more suggestions with respect to notebook ?

**Review by zoq — APPROVED** [review]
No further comments from my side, thanks for another great example.

**Review by kartikdutt18 — APPROVED** [review]
LGTM! Thanks for the contribution.
